File: docs/js/url-service.js
/**
 * URL Service for Frontend
 * Provides environment-aware URL generation for JavaScript components
 */

class URLService {
    constructor() {
        this.config = null;
        this.baseUrl = null;
        this.deploymentType = null;
        
        // Initialize from deployment config
        this.initializeFromConfig();
    }
    
    /**
     * Initialize URL service from deployment configuration
     */
    async initializeFromConfig() {
        try {
            // Try to load deployment config
            const configResponse = await fetch('./deployment-config.json');
            if (configResponse.ok) {
                this.config = await configResponse.json();
                this.baseUrl = this.config.baseUrl;
                this.deploymentType = this.config.deploymentType;
                console.log('📐 URL Service initialized:', this.config);
            } else {
                throw new Error('Config not found');
            }
        } catch (error) {
            // Fallback to detecting from current URL
            this.initializeFromCurrentURL();
        }
    }
    
    /**
     * Fallback initialization from current URL
     */
    initializeFromCurrentURL() {
        const currentUrl = window.location.href;
        const currentHost = window.location.host;
        const currentPath = window.location.pathname;
        
        // Detect if we're in a preview deployment
        if (currentPath.includes('/branch-preview/')) {
            // Extract branch preview path
            const match = currentPath.match(/^(.*?)\/branch-preview\/([^\/]+)/);
            if (match) {
                this.baseUrl = `${window.location.protocol}//${currentHost}${match[1]}/branch-preview/${match[2]}`;
                this.deploymentType = 'preview';
            }
        } else {
            // Assume production deployment
            this.baseUrl = `${window.location.protocol}//${currentHost}`;
            this.deploymentType = 'production';
        }
        
        this.config = {
            baseUrl: this.baseUrl,
            deploymentType: this.deploymentType,
            urls: {
                schemas: `${this.baseUrl}/schema/`,
                libraries: `${this.baseUrl}/libraries/`,
                widgets: `${this.baseUrl}/widgets/`,
                notebooks: `${this.baseUrl}/notebooks/`
            }
        };
        
        console.log('📐 URL Service fallback initialized:', this.config);
    }
    
    /**
     * Get base URL for current deployment
     */
    getBaseUrl() {
        return this.baseUrl || window.location.origin;
    }
    
    /**
     * Get URL for a schema file
     */
    getSchemaUrl(schemaPath) {
        const baseUrl = this.getBaseUrl();
        const cleanPath = schemaPath.replace(/^\/+/, '');
        return `${baseUrl}/${cleanPath}`;
    }
    
    /**
     * Get URL for a library resource
     */
    getLibraryUrl(libraryName, filePath = '') {
        const baseUrl = this.getBaseUrl();
        const cleanFilePath = filePath.replace(/^\/+/, '');
        if (cleanFilePath) {
            return `${baseUrl}/libraries/${libraryName}/${cleanFilePath}`;
        } else {
            return `${baseUrl}/libraries/${libraryName}/`;
        }
    }
    
    /**
     * Get URL for a widget
     */
    getWidgetUrl(widgetId) {
        const baseUrl = this.getBaseUrl();
        return `${baseUrl}/widgets/${widgetId}/`;
    }
    
    /**
     * Get URL for a notebook
     */
    getNotebookUrl(notebookPath) {
        const baseUrl = this.getBaseUrl();
        const cleanPath = notebookPath.replace(/^\/+/, '');
        return `${baseUrl}/notebooks/${cleanPath}`;
    }
    
    /**
     * Get download URL for notebooks.jsonld
     */
    getNotebooksJsonLdUrl() {
        return this.getSchemaUrl('notebooks.jsonld');
    }
    
    /**
     * Get deployment information
     */
    getDeploymentInfo() {
        return {
            baseUrl: this.baseUrl,
            deploymentType: this.deploymentType,
            config: this.config
        };
    }
    
    /**
     * Generate absolute URL from relative path
     */
    resolveUrl(relativePath) {
        const baseUrl = this.getBaseUrl();
        const cleanPath = relativePath.replace(/^\/+/, '');
        return `${baseUrl}/${cleanPath}`;
    }
    
    /**
     * Check if URL needs to be updated for current deployment
     */
    updateUrl(url) {
        if (!url) return url;
        
        // If it's already a relative URL, resolve it
        if (url.startsWith('./') || url.startsWith('../') || url.startsWith('/')) {
            return this.resolveUrl(url);
        }
        
        // If it's an absolute URL pointing to the same repository, update it
        const baseUrl = this.getBaseUrl();
        const repoMatch = url.match(/https:\/\/([^\/]+)\/([^\/]+)\/([^\/]+)(\/.*)?/);
        if (repoMatch) {
            const [, host, owner, repo, path] = repoMatch;
            if (host.includes('github.io') && path) {
                return `${baseUrl}${path}`;
            }
        }
        
        return url;
    }
}

// Create global URL service instance
window.urlService = new URLService();

// Export for module usage
if (typeof module !== 'undefined' && module.exports) {
    module.exports = URLService;
}

File: scripts/url_service.py
#!/usr/bin/env python3
"""
URL Service for GitHub Pages Deployment
Manages environment-aware URL generation for production and preview deployments
"""

import os
import json
import re
from typing import Dict, Optional, Any
from pathlib import Path


class URLService:
    """Service for generating deployment-aware URLs"""
    
    def __init__(self, repo_root: Optional[str] = None):
        self.repo_root = Path(repo_root) if repo_root else Path.cwd()
        
        # Get deployment configuration from environment variables
        self.github_repository = os.getenv('GITHUB_REPOSITORY', 'litlfred/notebooks')
        self.github_ref_name = os.getenv('GITHUB_REF_NAME', 'main')
        self.deployment_type = os.getenv('DEPLOYMENT_TYPE', 'preview')  # 'production' or 'preview'
        
        # Parse owner and repo from repository name
        parts = self.github_repository.split('/')
        self.owner = parts[0] if len(parts) > 0 else 'litlfred'
        self.repo = parts[1] if len(parts) > 1 else 'notebooks'
        
        # Generate base URL based on deployment type
        self.base_url = self._generate_base_url()
    
    def _generate_base_url(self) -> str:
        """Generate base URL based on deployment environment"""
        github_pages_base = f"https://{self.owner}.github.io/{self.repo}"
        
        if self.deployment_type == 'production':
            # Production deployment: {profile}.github.io/{repo}
            return github_pages_base
        else:
            # Preview deployment: {profile}.github.io/{repo}/branch-preview/{branch-name}
            safe_branch_name = self._sanitize_branch_name(self.github_ref_name)
            return f"{github_pages_base}/branch-preview/{safe_branch_name}"
    
    def _sanitize_branch_name(self, branch_name: str) -> str:
        """Sanitize branch name for URL usage"""
        # Replace invalid URL characters with hyphens
        sanitized = re.sub(r'[^a-zA-Z0-9\-_]', '-', branch_name)
        # Remove multiple consecutive hyphens
        sanitized = re.sub(r'-+', '-', sanitized)
        # Remove leading/trailing hyphens
        return sanitized.strip('-')
    
    def get_base_url(self) -> str:
        """Get the base URL for this deployment"""
        return self.base_url
    
    def get_schema_url(self, schema_path: str) -> str:
        """Generate URL for a schema file"""
        # Remove leading slash if present
        schema_path = schema_path.lstrip('/')
        return f"{self.base_url}/{schema_path}"
    
    def get_library_url(self, library_name: str, file_path: str = '') -> str:
        """Generate URL for a library resource"""
        file_path = file_path.lstrip('/')
        if file_path:
            return f"{self.base_url}/libraries/{library_name}/{file_path}"
        else:
            return f"{self.base_url}/libraries/{library_name}/"
    
    def get_widget_url(self, widget_id: str) -> str:
        """Generate URL for a widget"""
        return f"{self.base_url}/widgets/{widget_id}/"
    
    def get_deployment_info(self) -> Dict[str, Any]:
        """Get deployment information"""
        return {
            "deployment_type": self.deployment_type,
            "base_url": self.base_url,
            "repository": self.github_repository,
            "branch": self.github_ref_name,
            "owner": self.owner,
            "repo": self.repo,
            "sanitized_branch": self._sanitize_branch_name(self.github_ref_name)
        }
    
    def update_schema_urls(self, schema_content: str) -> str:
        """Update hardcoded URLs in schema content to be environment-aware"""
        # Replace hardcoded GitHub Pages URLs with relative or environment-aware URLs
        old_base = f"https://{self.owner}.github.io/{self.repo}"
        
        # Replace absolute URLs with relative paths where possible
        updated_content = schema_content.replace(old_base, self.base_url)
        
        return updated_content
    
    def generate_url_config(self) -> Dict[str, Any]:
        """Generate URL configuration for frontend"""
        return {
            "baseUrl": self.base_url,
            "deploymentType": self.deployment_type,
            "repository": self.github_repository,
            "branch": self.github_ref_name,
            "owner": self.owner,
            "repo": self.repo,
            "urls": {
                "schemas": f"{self.base_url}/schema/",
                "libraries": f"{self.base_url}/libraries/",
                "widgets": f"{self.base_url}/widgets/",
                "notebooks": f"{self.base_url}/notebooks/"
            }
        }
    
    def write_url_config(self, output_path: Optional[str] = None) -> str:
        """Write URL configuration to file"""
        if not output_path:
            output_path = self.repo_root / "docs" / "deployment-config.json"
        
        config = self.generate_url_config()
        
        output_file = Path(output_path)
        output_file.parent.mkdir(parents=True, exist_ok=True)
        
        with open(output_file, 'w') as f:
            json.dump(config, f, indent=2)
        
        print(f"✓ URL configuration written to: {output_file}")
        return str(output_file)


def main():
    """Main function for CLI usage"""
    import argparse
    
    parser = argparse.ArgumentParser(description='Generate URL configuration for deployment')
    parser.add_argument('--repo-root', default='.', help='Repository root directory')
    parser.add_argument('--output', help='Output file path')
    parser.add_argument('--deployment-type', choices=['production', 'preview'], 
                      help='Override deployment type')
    parser.add_argument('--branch', help='Override branch name')
    
    args = parser.parse_args()
    
    # Set environment variables if provided
    if args.deployment_type:
        os.environ['DEPLOYMENT_TYPE'] = args.deployment_type
    if args.branch:
        os.environ['GITHUB_REF_NAME'] = args.branch
    
    service = URLService(args.repo_root)
    
    print("🌐 URL Service Configuration")
    print("=" * 40)
    
    info = service.get_deployment_info()
    for key, value in info.items():
        print(f"  {key}: {value}")
    
    print("\n📋 Generated URLs:")
    config = service.generate_url_config()
    for category, url in config['urls'].items():
        print(f"  {category}: {url}")
    
    # Write configuration file
    config_file = service.write_url_config(args.output)
    
    print(f"\n✅ URL service initialized successfully")
    print(f"📁 Configuration: {config_file}")


if __name__ == "__main__":
    main()